### PR TITLE
Use Nerdbank.GitVersioning

### PR DESCRIPTION
**Pull request by sharwell**
~~⚠️ This pull request depends on #299, and will be substantially simplified once that PR is merged. ⚠️~~ 

## Overview

This change uses **Nerdbank.GitVersioning** to automatically version the assemblies and packages as work progresses in the repository.

## Deployment

Deploying a release to NuGet is much simpler after this change is applied. After setting up an environment (see next part), here are the steps to release a build:

1. After merging a pull request to master, locate the AppVeyor build you want to release
2. Open the **Image: Visual Studio 2017; Configuration: Release** job for the build (this is the only job containing the full set of correct artifacts for release)
3. At the top of the AppVeyor page, click **Deploy**
4. Select the **NuGet (coverlet)** environment and deploy
5. After deployment succeeds, create and push a tag in the repository

## Creating the environment

💡 This is a one-time setup process, after which you can use the environment any time you want to make a new coverlet release.

1. Log in to AppVeyor and go to the **Environments** tab (at the top, next to **Projects**)
1. Create a new environment
1. Select **NuGet** as the deployment provider
1. In your NuGet profile, create a new API key, and scope the key to `coverlet.*`
1. Back in AppVeyor environment setup apply the following:
    * Environment name: NuGet (coverlet)
    * API key: Copied from step (4)
1. Click **Add environment** to create the environment

### CODE REVIEW

**Comment by sharwell** [comment]
Marked as WIP while I rebase to account for recent merges.

**Comment by sharwell** [comment]
@tonerdo I left this marked as a WIP, but only because I want to rebase it after #299 is merged so the code review is simplified  to only show the changes in what is currently commit 948be428af3f1668fe6ba77aabf0a7748c05fe56.

**Comment by sharwell** [comment]
@tonerdo This is now rebased. As you can see, the builds produced by each AppVeyor build now have a unique version number, making it tremendously easier to prepare/manage releases. You will only rarely need to update version numbers in the project, and when you do, it will be limited to changes in **version.json**.

File: version.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "$schema": "https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json",
+  "version": "2.5",
+  "publicReleaseRefSpec": [
+    "^refs/heads/master$"
+  ],
+  "buildNumberOffset": 3,
```
Comment (sharwell) [inline version.json]: 📝 This accounts for the fact that you were already making releases with version numbers **2.5.x**. The property can be permanently removed starting with your first **2.6+** build.

**Comment by AArnott** [comment]
I'm torn. The doc looks well written, but it doesn't really have anything to do with NB.GV itself. But maybe as a new AppVeyor-specific suggestions doc linked from https://github.com/AArnott/Nerdbank.GitVersioning/blob/master/doc/cloudbuild.md ?

@sharwell would you be willing to send such a PR

**Comment by sharwell** [comment]
@tonerdo Conflicts are resolved again

**Comment by tonerdo** [comment]
Thank you @sharwell, will take a look in a couple

**Comment by tonerdo** [comment]
Superseded by #425

**Pull request closed without merging**